### PR TITLE
Add (Cluster)OpenIDConnectPreset resources

### LINKED ISSUE #1161: Introduce a new OpenID Connect configuration resource(s)
# Introduce new OpenID Connect configuration resource(s)

Gardener offers clusters with OpenID Connect (OIDC) not configured by default.

Offering a way to (optionaly) inject OIDC configuration on `Shoot` on creation can dramatically improve the secutity, and it can be used to limit the need for `kubeconfigs` with client x509 certificates and the admin (`system:masters`) certificate.

The solution should be:

- 100% backwards compatible.
- 100% customizable by cluster administrators on `cluster-scope`.
- 100% customizable by project administrators on `project-scope`.
- 100% side-effects free.
- `Shoot` configuration should be mutated on `CREATE` and not on `UPDATE`/`DELETE` (this behavior can be achieved with an additional (admission)controller, but not a requirement).

> Note: This is NOT a policy designed to enforce OpenID Configurations.

## Implementation

Two new API resources should be added to the `Gardener` - `OpenIDConnectConfiguration` and `ClusterOpenIDConnectConfiguration`

### OpenIDConnectConfiguration

`OpenIDConnectConfiguration` is a namespaced resource which holds a temple of OpenID Connect configuration, `Shoot` selector, and weight:

```yaml
apiVersion: settings.gardener.cloud/v1alpha1
kind: OpenIDConnectConfiguration
metadata:
  name: default
  namespace: some-namespace
spec:
  selector: {} # selects every Shoot in the default namespace.
  weight: 1 # Weight associated with matching the Shoot selector, in the range 1-100.
  config:
    caBundle: |
      -----BEGIN CERTIFICATE-----
      Li4u
      -----END CERTIFICATE-----
    clientID: client-id
    clientSecret: some-secret
    groupsClaim: groups-claim
    groupsPrefix: groups-prefix
    issuerURL: https://identity.example.com
    usernameClaim: username-claim
    usernamePrefix: username-prefix
    signingAlgs: RS256,some-other-algorithm
```

Fields in this spec include:

- `selector` a standard [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors), selecting to which `Shoot` this configuration is applied to.
- `weight` a value in the `1-100` range, controlling priority for selecing `OpenIDConnectConfiguration` for the same `Shoot`.
- `config` a template for `Shoot`'s `spec.kubernetes.kubeAPIServer.oidcConfig`.

### ClusterOpenIDConnectConfiguration

`ClusterOpenIDConnectConfiguration` is the cluster-scoped version of `OpenIDConnectConfiguration`, normally reserverd for cluster administrators:

```yaml
apiVersion: settings.gardener.cloud/v1alpha1
kind: ClusterOpenIDConnectConfiguration
metadata:
  name: default
spec:
  selector: {} # selects every Shoot in matched namespaces.
  namespaceSelector: {} # selects every namespace.
  weight: 1 # Weight associated with matching the Shoot selector, in the range 1-100.
  config:
    caBundle: |
      -----BEGIN CERTIFICATE-----
      Li4u
      -----END CERTIFICATE-----
    clientID: client-id
    clientSecret: some-secret
    groupsClaim: groups-claim
    groupsPrefix: groups-prefix
    issuerURL: https://identity.example.com
    usernameClaim: username-claim
    usernamePrefix: username-prefix
    signingAlgs: RS256,some-other-algorithm
```

Fields in this spec are the same as in `OpenIDConnectConfiguration`, but also include:

- `namespaceSelector` a standard [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors), selecting `Namespaces/Projects` to which this configuration is applied to.

### Garden API server admission controller

A new mutating admission controller should be added and enabled by default when a new `Shoot` admission request is recieved it should:

- Work only `CREATE` calls
- Mutate only Shoots without `spec.kubernetes.kubeAPIServer.oidcConfig`.
- Mutate `spec.kubernetes.kubeAPIServer.oidcConfig` with correct configuration.

## Ordering

When `Shoot` clusters are matched by multiple configuration the following ordering should take place:

1. `OpenIDConnectConfiguration` is selected over `ClusterOpenIDConnectConfiguration`.
1. `(Cluster)OpenIDConnectConfiguration` with the highest `spec.weight` is selected..
1. The first `(Cluster)OpenIDConnectConfiguration`(orderd by name) is selected.

## About clientSecret

While not needed / configurable on the API server, in order for the [kubectl's OIDC Authenticator](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#option-1-oidc-authenticator) to work `clientSecret` must be known for generating `kubeconfigs`.

This new optional field shall be added to the [OpenIDConfig](https://github.com/gardener/gardener/blob/5ab1e291b3cd4fb0a10346c73e39d649832b9e94/pkg/apis/garden/v1beta1/types.go#L1279-L1310) spec.

In both `ClusterOpenIDConnectConfiguration` and `OpenIDConnectConfiguration` this value should be required.

## Additional benefits

This new resources can also be used in https://github.com/gardener/dashboard/issues/419.

## Unknowns

- Wouldn't it make more sense to expose the entire [KubeAPIServer configuration](https://github.com/gardener/gardener/blob/5ab1e291b3cd4fb0a10346c73e39d649832b9e94/pkg/apis/garden/v1beta1/types.go#L1246-L1262) as a standalone resource?

### PR DESCRIPTION
**Pull request by mvladev**
**What this PR does / why we need it**:

This PR adds the following:

- new `settings.gardener.cloud` API group
- `OpenIDConnectPreset` and `ClusterOpenIDConnectPreset`
- New fields in the `Shoot` specification allowing for generation of kubeconfigs with `OpenIDConnect` settings in them
- `OpenIDConnectPreset` and `ClusterOpenIDConnectPreset` mutation admission controllers which apply those Presets to `Shoot` objects on `CREATE` requests.

**Which issue(s) this PR fixes**:
Fixes #1161 

**Special notes for your reviewer**:

**Release note**:

```noteworthy operator
The new `ClusterOpenIDConnectPreset` resource allows for specifying OpenID Connect configurations which are applied to `Projects` and `Shoot` cluster-wide.
```

```noteworthy operator
Two new mutation admission controllers are introduced - `ClusterOpenIDConnectPreset`  and `OpenIDConnectPreset`. Those controllers can are enabled by default and can be disabled with `--disable-admission-plugins` flag on gardener-apiserver.
```

```noteworthy user
The new `OpenIDConnectPreset` resource allows for specifying OpenID Connect configurations which are applied to `Shoot` namespace-wide.
```

```noteworthy user
A new optional field in `Shoot` specification has been added `spec.kubernetes.kubeAPIServer.oidcConfig.clientAuthentication`. It can specify OpenID Connect settings used for `kubeconfig` generation.
```

```noteworthy developer
`./hack/dev-setup-register-gardener` must be ran to register the new `settings.gardener.cloud` API group.
```
/cc @vasu1124

### CODE REVIEW

**Comment by mvladev** [comment]
ping

**Comment by rfranzke** [comment]
I’ll review it on Monday, I need to finish the Shoot API Migration this week. Sorry for the delay.

File: docs/concepts/openidconenct-presets.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
+# ClusterOpenIDConnectPreset and OpenIDConenctPreset
+
+This page provides an overview of ClusterOpenIDConnectPresets and OpenIDConenctPresets, which are objects for injecting [OpenIDConnect Configuration](https://openid.net/connect/) into `Shoot` at creation time. The injected information contains configuration for the Kube API Server and optionally configuration for kubeconfig generation using said configuration.
+
+## OpneIDConnectPresets
+
+An OpenIDConnectPresets is an API resource for injecting additional runtime OIDC requirements into a Shoot at creation time. You use label selectors to specify the `Shoot` to which a given OpneIDCOnenctPreset applies.
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
An OpenIDConnectPreset is an API resource for injecting additional runtime OIDC requirements into a Shoot at creation time. You can use label selectors to specify the `Shoot` to which a given OpenIDConnectPresets applies.
```

File: docs/concepts/openidconenct-presets.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
+# ClusterOpenIDConnectPreset and OpenIDConenctPreset
+
+This page provides an overview of ClusterOpenIDConnectPresets and OpenIDConenctPresets, which are objects for injecting [OpenIDConnect Configuration](https://openid.net/connect/) into `Shoot` at creation time. The injected information contains configuration for the Kube API Server and optionally configuration for kubeconfig generation using said configuration.
+
+## OpneIDConnectPresets
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
## OpenIDConnectPreset
```

File: docs/concepts/openidconenct-presets.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+An OpenIDConnectPresets is an API resource for injecting additional runtime OIDC requirements into a Shoot at creation time. You use label selectors to specify the `Shoot` to which a given OpneIDCOnenctPreset applies.
+
+Using a OpenIDConnectPresets allows project owners to not have to explicitly provide the same OIDC configuration for every `Shoot` in their `Project`.
+
+For more information about the background, see the [issue](https://github.com/gardener/gardener/issues/1161) for OpenIDConnectPreset.
+
+### How OpenIDConnectPresets works
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
### How OpenIDConnectPreset works
```

File: docs/concepts/openidconenct-presets.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+- If multiple presets are matched then only one is chosen and results are sorted based on:
+
+  1. `.weight` value.
+  1. lexicographically ordering their names ( e.g. `002preset` > `001preset` )
+
+- If the `Shoot` already has a `.spec.kubernetes.kubeAPIServer.oidcConfig` then no mutation occures.
+
+### Simple OpenIDConnectPreset example
+
+This is a siople example to show how a `Shoot` is modified by the OpenIDConnectPreset
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
This is a simple example to show how a `Shoot` is modified by the OpenIDConnectPreset
```

File: docs/concepts/openidconenct-presets.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+Create the OpenIDConnectPreset:
+
+```console
+kubectl apply -f preset.yaml
+```
+
+Examine the created OpenIDConnectPreset:
+
+```bash
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
```console
```
to make it the same as the other usages in the same file.

File: docs/concepts/openidconenct-presets.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+kubectl get openidconnectpresets
+NAME     ISSUER            SHOOT-SELECTOR   AGE
+test-1   https://foo.bar   oidc=enabled     1s
+```
+
+Simple `Shoot` example:
+
+This is a sample of a `Shoot` with some fields ommited:
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]: You could remove one of the lines, they both say that an example is coming.

File: docs/concepts/openidconenct-presets.md [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+For example:
+
+```text
+--disable-admission-plugins=OpenIDConnectPreset
+```
+
+## ClusterOpenIDConnectPresets
+
+A ClusterOpenIDConnectPresets is an API resource for injecting additional runtime OIDC requirements into a Shoot at creation time. In contrast to OpenIDConenct it's a cluster-scoped resource. You use label selectors to specify the `Project` and `Shoot` to which a given OpneIDCOnenctPreset applies.
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
A ClusterOpenIDConnectPreset is an API resource for injecting additional runtime OIDC requirements into a Shoot at creation time. In contrast to OpenIDConnectPreset it's a cluster-scoped resource. You can use label selectors to specify the `Project` and `Shoot` to which a given ClusterOpenIDConnectPreset applies.
```

File: docs/concepts/openidconenct-presets.md [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+The OpenIDConnectPreset admission control is enabled by default. To disable it use the `--disable-admission-plugins` flag on the gardener-apisever.
+
+For example:
+
+```text
+--disable-admission-plugins=OpenIDConnectPreset
+```
+
+## ClusterOpenIDConnectPresets
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
## ClusterOpenIDConnectPreset
```

File: docs/concepts/openidconenct-presets.md [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+Using a OpenIDConnectPresets allows cluster owners to not have to explicitly provide the same OIDC configuration for every `Shoot` in specific `Project`.
+
+For more information about the background, see the [issue](https://github.com/gardener/gardener/issues/1161) for ClusterOpenIDConnectPreset.
+
+### How ClusterOpenIDConnectPresets works
+
+Gardener provides an admission controller (ClusterOpenIDConnectPreset) which, when enabled, applies ClusterOpenIDConnectPresets to incoming `Shoot` creation requests. When a `Shoot` creation request occurs, the system does the following:
+
+- Retrieve all ClusterOpenIDConnectPresets available.
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]: You could use ordered list to indicate that the steps are executed sequentially in the given order.

File: docs/concepts/openidconenct-presets.md [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+- Check if the project label selector of any ClusterOpenIDConnectPreset matches the labels of the `Project` in which the `Shoot` is beign created.
+- Check if the shoot label selectors of any ClusterOpenIDConnectPreset matches the labels on the `Shoot` being created.
+- If multiple presets are matched then only one is chosen and results are sorted based on:
+
+  1. `.weight` value.
+  1. lexicographically ordering their names ( e.g. `002preset` > `001preset` )
+
+- If the `Shoot` already has a `.spec.kubernetes.kubeAPIServer.oidcConfig` then no mutation occures.
+
+> Note: Due to the previous requiremenet if a `Shoot` is matched by both `OpenIDConnectPreset` and `ClusterOpenIDConnectPreset` then `OpenIDConnectPreset` takes precidence over `ClusterOpenIDConnectPreset`.
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
> Note: Due to the previous requirement if a `Shoot` is matched by both `OpenIDConnectPreset` and `ClusterOpenIDConnectPreset` then `OpenIDConnectPreset` takes precedence over `ClusterOpenIDConnectPreset`.
```

File: example/10-clusteropenidconnectpreset.yaml [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# Plant cluster registration manifest through which an external kubernetes cluster will be mapped
```
Comment (ialidzhikov) [inline example/10-clusteropenidconnectpreset.yaml]: You could update this comment as it is not for ClusterOpenIDConnectPreset.

File: docs/concepts/openidconenct-presets.md [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+### Simple ClusterOpenIDConnectPreset example
+
+This is a siople example to show how a `Shoot` is modified by the ClusterOpenIDConnectPreset
+
+```yaml
+apiVersion: settings.gardener.cloud/v1alpha1
+kind: ClusterOpenIDConnectPreset
+metadata:
+  name:  test
+shootSelector:
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]: Not sure why we don't have `.spec` field? I guess this is the convention.
But however you have more experience with the API conventions and you could share what was your motivation.

File: example/10-clusteropenidconnectpreset.yaml [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+---
+apiVersion: settings.gardener.cloud/v1alpha1
+kind: ClusterOpenIDConnectPreset
+metadata:
+  name:  example-preset
+  namespace: garden-dev
```
Comment (ialidzhikov) [inline example/10-clusteropenidconnectpreset.yaml]: You could remove the ns as it is cluster-scoped.

File: example/10-openidconnectpreset.yaml [hunk 14]
```diff
@@ -0,0 +1,32 @@
+# Plant cluster registration manifest through which an external kubernetes cluster will be mapped
```
Comment (ialidzhikov) [inline example/10-openidconnectpreset.yaml]: You could update this comment as it is not for OpenIDConnectPreset.

File: docs/concepts/openidconenct-presets.md [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+  1. lexicographically ordering their names ( e.g. `002preset` > `001preset` )
+
+- If the `Shoot` already has a `.spec.kubernetes.kubeAPIServer.oidcConfig` then no mutation occures.
+
+### Simple OpenIDConnectPreset example
+
+This is a siople example to show how a `Shoot` is modified by the OpenIDConnectPreset
+
+```yaml
+apiVersion: settings.gardener.cloud/v1alpha1
```
Comment (ialidzhikov) [inline docs/concepts/openidconenct-presets.md]: Isn't `settings.gardener.cloud/v1alpha1` too generic? What about `authorization.gardener.cloud/v1alpha1`/`authentication.gardener.cloud/v1alpha1`?

File: pkg/apis/settings/v1alpha1/defaults_test.go [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+	})
+
+})
+
+func defaultSpec() v1alpha1.OpenIDConnectPresetSpec {
+
+	return v1alpha1.OpenIDConnectPresetSpec{
+		Server: v1alpha1.KubeAPIServerOpenIDConnect{
+			// string literal are used to be sure that the test fails
```
Comment (ialidzhikov) [inline pkg/apis/settings/v1alpha1/defaults_test.go:73]:
```golang
        usernameClaim := "sub"
	return v1alpha1.OpenIDConnectPresetSpec{
		Server: v1alpha1.KubeAPIServerOpenIDConnect{
			// string literal are used to be sure that the test fails
			// if the constant values are changed.
			UsernameClaim: &usernameClaim,
			SigningAlgs:   []string{"RS256"},
		},
		ShootSelector: &metav1.LabelSelector{},
	}
```

and you can remove the usage of `k8s.io/utils/pointer`.

**Review by ialidzhikov — CHANGES_REQUESTED** [review]
I will appreciate if you rebase. :)

File: docs/concepts/openidconenct-presets.md [hunk 17]
```diff
@@ -0,0 +1,288 @@
+# ClusterOpenIDConnectPreset and OpenIDConenctPreset
```
Comment (rfranzke) [inline docs/concepts/openidconenct-presets.md]:
```suggestion
# ClusterOpenIDConnectPreset and OpenIDConnectPreset
```

+ similar typo in filename itself

File: docs/concepts/openidconenct-presets.md [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+NAME     ISSUER            SHOOT-SELECTOR   AGE
+test-1   https://foo.bar   oidc=enabled     1s
+```
+
+Simple `Shoot` example:
+
+This is a sample of a `Shoot` with some fields ommited:
+
+```yaml
+kind: Shoot
```
Comment (rfranzke) [inline docs/concepts/openidconenct-presets.md]: Swap `kind`/`apiVersion`

File: pkg/apis/settings/types_shared.go [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+	metav1.ObjectMetaAccessor
+	GetPresetSpec() *OpenIDConnectPresetSpec
+	SetPresetSpec(s *OpenIDConnectPresetSpec)
+}
+
+var _ Preset = &OpenIDConnectPreset{}
+var _ Preset = &ClusterOpenIDConnectPreset{}
+
+// GetPresetSpec returns a pointer to the OpenIDConnect specification.
+func (o *OpenIDConnectPreset) GetPresetSpec() *OpenIDConnectPresetSpec {
```
Comment (rfranzke) [inline pkg/apis/settings/types_shared.go]: implement these in `types_openidconnect_preset.go` and `types_cluster_....go`?

File: pkg/apis/settings/v1alpha1/defaults_test.go [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestAPI(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "v1alpha1 Suite")
```
Comment (rfranzke) [inline pkg/apis/settings/v1alpha1/defaults_test.go]:
```suggestion
	RunSpecs(t, "Settings v1alpha1 Suite")
```

File: pkg/registry/garden/rest/storage_garden.go [hunk 21]
```diff
@@ -69,7 +69,6 @@ func (p StorageProvider) v1beta1Storage(restOptionsGetter generic.RESTOptionsGet
 	seedStorage := seedstore.NewStorage(restOptionsGetter, cloudprofileStorage.CloudProfile)
 	storage["seeds"] = seedStorage.Seed
 	storage["seeds/status"] = seedStorage.Status
-
```
Comment (rfranzke) [inline pkg/registry/garden/rest/storage_garden.go]: ?

File: pkg/registry/settings/clusteropenidconnectpreset/storage/tableconvertor.go [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+
+		cells = append(cells, obj.Name)
+
+		if issuer := obj.Server.IssuerURL; len(issuer) > 0 {
+			cells = append(cells, issuer)
+		} else {
+			cells = append(cells, "<unknown>")
+		}
+
+		cells = append(cells, metav1.FormatLabelSelector(obj.ProjectSelector), metatable.ConvertToHumanReadableDateType(obj.CreationTimestamp))
```
Comment (rfranzke) [inline pkg/registry/settings/clusteropenidconnectpreset/storage/tableconvertor.go]: Why is the creation timestamp appended twice?

File: pkg/apis/settings/validation/cluster_openidconnect_preset.go [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ValidateClusterOpenIDConnectPreset validates a ClusterOpenIDConnectPreset object.
+func ValidateClusterOpenIDConnectPreset(oidc *settings.ClusterOpenIDConnectPreset) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateObjectMeta(&oidc.ObjectMeta, false, apivalidation.NameIsDNSLabel, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, validateOpenIDConnectPresetSpec(&oidc.OpenIDConnectPresetSpec)...)
+
```
Comment (rfranzke) [inline pkg/apis/settings/validation/cluster_openidconnect_preset.go]: Can you please also call `metav1validation.ValidateLabelSelector()`?

File: pkg/apis/settings/validation/openidconnect_preset.go [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ValidateOpenIDConnectPreset validates a OpenIDConnectPreset object.
+func ValidateOpenIDConnectPreset(oidc *settings.OpenIDConnectPreset) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateObjectMeta(&oidc.ObjectMeta, true, apivalidation.NameIsDNSLabel, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, validateOpenIDConnectPresetSpec(&oidc.OpenIDConnectPresetSpec)...)
+
```
Comment (rfranzke) [inline pkg/apis/settings/validation/openidconnect_preset.go:29]: Can you please also call `metav1validation.ValidateLabelSelector()`?

File: plugin/pkg/shoot/oidc/openidconnectpreset/admission.go [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+	// We have an OpenIDConnectPreset, use it.
+	if preset != nil {
+		applier.ApplyOIDCConfiguration(shoot, preset)
+		return nil
+	}
+
+	return nil
+}
+
+func filterOIDCs(oidcs []*settingsv1alpha1.OpenIDConnectPreset, shoot *garden.Shoot) (*settingsv1alpha1.OpenIDConnectPresetSpec, error) {
```
Comment (rfranzke) [inline plugin/pkg/shoot/oidc/openidconnectpreset/admission.go:142]: Can this be a utility function used by both the `OpenIDConnectPreset` and `ClusterOpenIDConnectPreset` admission plugins?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Very nice PR, thanks!

Apart from the inline suggestions, here a few more things:

1. According to my understanding `ClusterOpenIDConnectPreset` resources should only be set by operators/administrators, but not by end-users. End-users shall only use `OpenIDConnectPreset` resources in their project namespace(s). --> Please add the required RBAC roles to the https://github.com/gardener/gardener/blob/master/charts/gardener/charts/application/templates/rbac-user.yaml file.
1. I don't see that the `clientConnection` field in the OIDC config structure is used anywhere. Do I miss something?
1. Please add a release note that developers have to run `./hack/dev-setup-register-gardener` to register the new `settings.gardener.cloud` API group.

File: pkg/apis/settings/types_shared.go (hunk 19 above)
Comment (mvladev) [reply pkg/apis/settings/types_shared.go]: Those are used only for validation testing.

Comment (rfranzke) [reply pkg/apis/settings/types_shared.go]: ? I don't understand, you have dedicated `types_` files which suggest that everything belonging to this type should be in this file, but here you don't follow this approach.

Comment (mvladev) [reply pkg/apis/settings/types_shared.go]: Ahh, I thought you meant that this was missing in the versioned group. I'll fix it.

File: pkg/registry/garden/rest/storage_garden.go (hunk 21 above)
Comment (mvladev) [reply pkg/registry/garden/rest/storage_garden.go]: removed by mistake

File: plugin/pkg/shoot/oidc/openidconnectpreset/admission.go (hunk 25 above)
Comment (mvladev) [reply plugin/pkg/shoot/oidc/openidconnectpreset/admission.go:142]: I've tried but I couldn't come up with something good enough. No generics...

Comment (rfranzke) [reply plugin/pkg/shoot/oidc/openidconnectpreset/admission.go:142]: Ok

**Comment by mvladev** [comment]
> According to my understanding ClusterOpenIDConnectPreset resources should only be set by operators/administrators, but not by end-users. End-users shall only use OpenIDConnectPreset resources in their project namespace(s). --> Please add the required RBAC roles to the https://github.com/gardener/gardener/blob/master/charts/gardener/charts/application/templates/rbac-user.yaml file.

Fixed in be33f37 

> I don't see that the clientConnection field in the OIDC config structure is used anywhere. Do I miss something?

I'll add openid connect kubeconfig generation later on

> Please add a release note that developers have to run ./hack/dev-setup-register-gardener to register the new settings.gardener.cloud API group.

Done.

@ialidzhikov 

I'll add the specification field

**Comment by rfranzke** [comment]
> I'll add openid connect kubeconfig generation later on

ok, can you please create an issue and assign yourself?

**Comment by mvladev** [comment]
@ialidzhikov Cluster/OpenIDConnectProviderPreset now have a specification. 8e8440e 
@rfranzke I've create a #1433 issue for this.

**Comment by rfranzke** [comment]
Why did you rename the API group? I actually liked `settings.gardener.cloud`. @mandelsoft asked, once you have completed the first implementation of this API group, to introduce a `DNSProvider` resource in `settings.gardener.cloud` where end-users/project-owners may define standard DNS settings (similar to the standard OIDC settings you implemented). Now, with `authentication.gardener.cloud` this doesn't fit anymore and the concept cannot be extended without tremendous effort (of introducing a new API group again).

**Comment by mvladev** [comment]
I'll revert the change group commit.

File: docs/concepts/openidconenct-presets.md (hunk 15 above)
Comment (mvladev) [reply docs/concepts/openidconenct-presets.md]: See https://github.com/gardener/gardener/pull/1394#issuecomment-532172247

**Comment by mvladev** [comment]
Okay everything is back to `settings.gardener.cloud`

**Review by rfranzke — DISMISSED** [review]

**Comment by rfranzke** [comment]
@ialidzhikov can you please take a second look so that we can get this merged?

**Comment by ialidzhikov** [comment]
> @ialidzhikov can you please take a second look so that we can get this merged?

I am on it. Sorry for the delay.

File: docs/concepts/openidconnect-presets.md [hunk 26]
```diff
@@ -0,0 +1,288 @@
+# ClusterOpenIDConnectPreset and OpenIDConnectPreset
+
+This page provides an overview of ClusterOpenIDConnectPresets and OpenIDConnectPresets, which are objects for injecting [OpenIDConnect Configuration](https://openid.net/connect/) into `Shoot` at creation time. The injected information contains configuration for the Kube API Server and optionally configuration for kubeconfig generation using said configuration.
+
+## OpneIDConnectPreset
```
Comment (ialidzhikov) [inline docs/concepts/openidconnect-presets.md]:
```suggestion
## OpenIDConnectPreset
```

File: docs/concepts/openidconnect-presets.md [hunk 27]
```diff
@@ -0,0 +1,288 @@
...
+  1. `.spec.weight` value.
+  1. lexicographically ordering their names ( e.g. `002preset` > `001preset` )
+
+- If the `Shoot` already has a `.spec.kubernetes.kubeAPIServer.oidcConfig` then no mutation occures.
+
+> Note: Due to the previous requirement if a `Shoot` is matched by both `OpenIDConnectPreset` and `ClusterOpenIDConnectPreset` then `OpenIDConnectPreset` takes precidence over `ClusterOpenIDConnectPreset`.
+
+### Simple ClusterOpenIDConnectPreset example
+
+This is a siople example to show how a `Shoot` is modified by the ClusterOpenIDConnectPreset
```
Comment (ialidzhikov) [inline docs/concepts/openidconnect-presets.md]:
```suggestion
This is a simple example to show how a `Shoot` is modified by the ClusterOpenIDConnectPreset
```

**Review by ialidzhikov — COMMENTED** [review]
Minor nits.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]

**Pull request merged**